src/Streamly/Internal/Data/Unfold/Enumeration.hs

```diff
@@ -0,0 +1,366 @@
+--
+-- License     : BSD3
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- The functions defined in this module should be rarely needed for direct use,
+-- try to use the operations from the 'Enumerable' type class
+-- instances instead.
+--
+-- This module provides an 'Enumerable' type class to enumerate 'Enum' types
+-- into a stream. The operations in this type class correspond to similar
+-- perations in the 'Enum' type class, the only difference is that they produce
+-- a stream instead of a list. These operations cannot be defined generically
+-- based on the 'Enum' type class. We provide instances for commonly used
+-- types. If instances for other types are needed convenience functions defined
+-- in this module can be used to define them. Alternatively, these functions
+-- can be used directly.
+--
+module Streamly.Internal.Data.Unfold.Enumeration
+(
+   -- * Enumerations
+     Enumerable (..)
+   -- ** Enumerate Num
+   , enumerateFromStepNum
+   , numFrom
+
+  -- ** Enumerating 'Bounded' 'Integral' Types
+  , enumerateFromIntegral
+  , enumerateFromThenIntegral
+
+  -- ** Enumerating 'Integral' Types
+  , enumerateFromToIntegral
+  , enumerateFromThenToIntegral
+
+  -- ** Enumerating unbounded 'Integral' Types
+  , enumerateFromStepIntegral
+
+  -- ** Enumerating 'Fractional' Types
+  , enumerateFromFractional
+  , enumerateFromToFractional
+  , enumerateFromThenFractional
+  , enumerateFromThenToFractional
+  )
+where
+
+import Data.Fixed
+import Data.Int
+import Data.Ratio
+import Data.Word
+import Numeric.Natural
+import Data.Functor.Identity (Identity(..))
+import Streamly.Internal.Data.Unfold.Type
+import Streamly.Internal.Data.Stream.StreamD.Step (Step(..))
+import Prelude
+       hiding (map, mapM, takeWhile, take, filter, const, zipWith
+              , drop, dropWhile)
+
+-------------------------------------------------------------------------------
+-- Utilities ------------------------------------------------------------------
+-------------------------------------------------------------------------------
+{-# INLINE_NORMAL supplySecond #-}
+supplySecond :: b -> Unfold m (a, b) c -> Unfold m a c
+supplySecond b = lmap (, b)
+
+{-# INLINE_NORMAL takeWhileM #-}
+takeWhileM :: Monad m => (b -> m Bool) -> Unfold m a b -> Unfold m a b
+takeWhileM f (Unfold step1 inject1) = Unfold step inject1
+  where
+    {-# INLINE_LATE step #-}
+    step st = do
+        r <- step1 st
+        case r of
+            Yield x s -> do
+                b <- f x
+                return $ if b then Yield x s else Stop
+            Skip s -> return $ Skip s
+            Stop   -> return Stop
+
+-- | End the stream generated by the 'Unfold' as soon as the predicate fails
+-- on an element.
+--
+-- /Since: 0.8.0/
+--
+{-# INLINE takeWhile #-}
+takeWhile :: Monad m => (b -> Bool) -> Unfold m a b -> Unfold m a b
+takeWhile f = takeWhileM (return . f)
+
```

Comment:
takeWhile etc and supplySecond should not be duplicated in the Enumeration module. Those can be moved from Unfold module to Unfold/Type module and you can import Unfold/Type in Enumeration module.